### ISSUE TITLE
[<Ray component: Core|RLlib|etc...>] Action masking not working properly on DQN

### ISSUE THREAD

**Issue by GuillermoHijano** [ray-project/ray#47537]
### What happened + What you expected to happen

Hi, I wrote down the code of a simple experiment that uses action masking for DQN to show that the action masking does not work properly. I have a simple custom environment, and I handle the action masking in the forward() method of the custom model. If I don't apply action masking (using `return action_scores, state` in forward()) I see that the trained agent performs games as expected (test provides non zero reward episodes). However, if I do apply action masking (using `return action_scores + inf_mask, state` to introduce very large negative numbers for the q-values of invalid actions as suggested in the documentation) the training is not successful anymore. The trained agent will make dumb decisions. 

### Versions / Dependencies

I am using:
Python 3.12.4
Ray 2.34.0

### Reproduction script

Here is the code of the experiment:

`import gymnasium as gym 
import numpy as np 
from gymnasium import spaces
import tqdm
import pandas as pd
import random 

from typing import Optional

NumStates=10
NumActions=4

def get_valid_actions(state):
    if state==9:
        return np.array([1,1,0,1])
    elif state==0:
        return np.array([1,1,1,0])
    else:
        return np.array([1,1,1,1])

class CustomDiscreteEnv(gym.Env):
    def __init__(self):
        super(CustomDiscreteEnv, self).__init__()
        self.action_space = spaces.Discrete(NumActions)  # Example: 4 possible actions
        self.observation_space = spaces.Dict(
                    {
                        "observations": spaces.Discrete(NumStates),  # Example: 10 possible states
                        "action_mask": spaces.Box(0,1,shape=(NumActions,), dtype=int)#spaces.MultiDiscrete([2]*NumActions)
                    }
                )
        self.state = 0

    def reset(self, *, seed=None, options: Optional[int] = None):
        self.state = self.observation_space["observations"].sample()
        self.steps=0
        return {"observations":np.array(self.state).astype(np.uint8),"action_mask":get_valid_actions(self.state)}, {}

    def step(self, action):
        self.steps+=1

        if action==2 and self.state<9:
            self.state+=1
        elif action==3 and self.state>0:
            self.state-=1
        
        reward = 1 if self.state == 7 else 0  # Example: reward for action 1
        done = self.steps==10
        info = {}
        truncated=False
        return {"observations":np.array(self.state).astype(np.uint8),"action_mask":get_valid_actions(self.state)}, reward, done, truncated, info

    def render(self, mode='human'):
        pass  # Implement rendering if needed

    def close(self):
        pass  # Implement closing if needed

from gymnasium.spaces import Box, Discrete
from ray.rllib.algorithms.dqn import DQNConfig
from ray.rllib.algorithms.dqn.dqn_torch_model import DQNTorchModel
from ray.rllib.models import ModelCatalog
from ray.rllib.models.torch.fcnet import FullyConnectedNetwork as TorchFC
from ray.rllib.utils.framework import try_import_torch
from ray.rllib.utils.torch_utils import FLOAT_MAX, FLOAT_MIN
torch, nn = try_import_torch()

class DQNActionMaskModel(DQNTorchModel):
    """PyTorch version of above ParametricActionsModel."""

    def __init__(
        self,
        obs_space,#: Box,
        action_space: Discrete,
        num_outputs,
        model_config,
        name,
        **kw,
    ):
        DQNTorchModel.__init__(
            self, obs_space, action_space, num_outputs, model_config, name, **kw
        )

        obs_len = obs_space.shape[0] - action_space.n
        orig_obs_space = Box(
            shape=(obs_len,), low=obs_space.low[:obs_len], high=obs_space.high[:obs_len]
        )
        self.action_embed_model = TorchFC(
            orig_obs_space,
            action_space,
            action_space.n,
            model_config,
            name + "_action_embed",
        )

    def forward(self, input_dict, state, seq_lens):
        action_mask = input_dict["obs"]["action_mask"]
        action_scores, _ = self.action_embed_model(
            {"obs": input_dict["obs"]["observations"]}
        )
        inf_mask =torch.where(action_mask > 0, 0.0, -1e8)
        #return action_scores + inf_mask, state
        return action_scores, state

    def value_function(self):
        return self.action_embed_model.value_function()
    
ModelCatalog.register_custom_model("custom_torch_model", DQNActionMaskModel)

import ray 
from ray.tune.registry import register_env
from ray.rllib.algorithms.dqn.dqn import DQNConfig
from scipy.special import softmax

def env_creator(env_config):
    env = CustomDiscreteEnv()
    return env 
register_env('custom_discrete_env', env_creator)

from ray.rllib.algorithms.dqn import DQN
from ray import air,train
from ray import tune
config = DQNConfig()
config = config.training(num_atoms=tune.grid_search([1,]),lr=tune.grid_search([0.0001]),hiddens=[],model={"custom_model": "custom_torch_model"})#TO_DO: Adding the hiddens=[] makes the model correctly receive the num_outputs parameter. Check if I introduce any issue by adding it
config = config.environment(env='custom_discrete_env')
config.timesteps_per_iteration=6000
results = tune.Tuner(
    "DQN",
    run_config=train.RunConfig(stop={"training_iteration":10},checkpoint_config=train.CheckpointConfig(checkpoint_at_end=True)),
    param_space=config.to_dict()
).fit()
best_result = results.get_best_result(
    metric="env_runners/episode_return_mean", mode="max"
)
best_checkpoint = best_result.checkpoint
algo=DQN.from_checkpoint(best_checkpoint)

#algo = config.build()
policy = algo.get_policy()
print(policy.model)

`

### Issue Severity

High: It blocks me from completing my task.
